### ISSUE TITLE
解剖一致性content_contrastive_loss

### ISSUE THREAD

**Issue by Mike-ihr** [CuCl-2/private_commit#8]
我在huggingface上提供的训练日志中发现content_contrastive_loss一直都维系在2.1408这个水平。请问这说明其实解剖特征并没有很好的拉近，对吗

**Comment by Mike-ihr** [comment]
<img width="1205" height="561" alt="Image" src="https://github.com/user-attachments/assets/4f076d2d-867d-4d75-9ab3-222aed5b5679" />这是我使用huggingface上提供的权重，在BraTs数据集上做的T-SNE可视化结果。我发现解剖特征并没有聚合在一起，与论文中report的并不一致

**Comment by CuCl-2** [comment]
https://github.com/CuCl-2/DC-Seg/issues/5#issuecomment-3817745526
看看这个，是不是用了crop_size 80, 导致和训练时不一致

**Comment by CuCl-2** [comment]
如果这个解决了你的问题和我说一下，我好update一下，避免大家老是遇到这个问题

**Comment by Mike-ihr** [comment]
应该不是这个原因。虽然我自己训练的时候确实使用了crop_size 80。但同时我也使用了huggingface上提供的权重，做这个测试的时候crop_size为120。但都得到同样的可视化结果。且在huggingface上提供的训练日志中，确实是content_contrastive_loss没有降低

**Comment by Mike-ihr** [comment]
<img width="2102" height="420" alt="Image" src="https://github.com/user-attachments/assets/bce800b4-31ad-4146-b1f0-8253f8e457fe" /> 日志里loss确实是不变的

**Comment by CuCl-2** [comment]
不是120呀，是112，你再试一下。loss是这样的，确实降低幅度很低，但从可视化效果看起来是可以的。
你再试一下112，如果还不行的话和我说一声

**Comment by Mike-ihr** [comment]
设置成112还是不行，我是在BraTS2018数据集上测试的。请问有相关的t-sne可视化脚本嘛

<img width="1375" height="563" alt="Image" src="https://github.com/user-attachments/assets/d2a4f33e-113c-4eed-87f9-cf128e5e99ce" />
`import os
import torch
import torch.nn.functional as F
import numpy as np
import matplotlib.pyplot as plt
from sklearn.manifold import TSNE
import warnings
warnings.filterwarnings("ignore")
os.environ["CUDA_LAUNCH_BLOCKING"] = "1"  
import models
from data.datasets_nii import Brats_loadall_test_nii
from torch.utils.data import DataLoader

def main():
    # ------------------ 配置区 ------------------
    datapath = '/root/data/2-MICCAI_BraTS_2018/processed_BRATS2018_DC_Seg'
    # 请根据你实际训练保存的权重路径修改下面这一行，如果还没有权重可以跑一个Epoch或使用下方的未经训练的初始网络权重测试能否正常出图
    checkpoint_path = '/root/dcreg/model_last.pth'  
    
    num_samples = 20     # 提取用于展示分布的测试集样本数量
    crop_size = 112       # 为了防止爆显存，提取中心 crop 尺寸

    # 1. 挂载模型
    print("Loading model...")
    model = models.DC_Seg(num_cls=4, fusion_type='RFM')
    model = torch.nn.DataParallel(model).cuda()
    
    if os.path.exists(checkpoint_path):
        checkpoint = torch.load(checkpoint_path)
        model.load_state_dict(checkpoint['state_dict'])
        print(f"Successfully loaded checkpoint from {checkpoint_path}")
    else:
        print(f"Warning: Checkpoint not found at {checkpoint_path}. Using initialized weights for testing.")

    # 2. 读取测试数据
    test_set = Brats_loadall_test_nii(transforms="Compose([NumpyType((np.float32, np.int64)),])", root=datapath, test_file='test3.txt')
    test_loader = DataLoader(dataset=test_set, batch_size=1, shuffle=False)

    # BraTS 有 4 种模态 (0:FLAIR, 1:T1ce, 2:T1, 3:T2)
    mask = torch.tensor([[True, True, True, True]]).cuda() 
    modality_names = ['FLAIR', 'T1c', 'T1', 'T2']

    all_contents = {0: [], 1: [], 2: [], 3: []}
    all_styles = {0: [], 1: [], 2: [], 3: []}

    print("Extracting features from network...")
    with torch.no_grad():
        for i, (x, target, name) in enumerate(test_loader):
            if i >= num_samples:
                break
            print(f"Processing sample {i+1}/{num_samples}: {name[0]}")
            
            # 使用中心裁剪防止显存溢出
            _, _, H, W, D = x.shape
            h_idx = (H - crop_size) // 2
            w_idx = (W - crop_size) // 2
            d_idx = (D - crop_size) // 2
            x_crop = x[:, :, h_idx:h_idx+crop_size, w_idx:w_idx+crop_size, d_idx:d_idx+crop_size].cuda()
            
            # 强制开启 is_training 标志 (在源码预测模式下它不返回 representations)
            model.module.is_training = True
            
            # 前向传播截取表示
            outputs = model(x_crop, mask)
            # outputs tuple 包含 contents 和 styles 如下：
            # fuse_pred, sep_preds, prm_preds, recon_out, mu_list, sigma_list, contents, styles
            contents = outputs[-2] 
            styles = outputs[-1]

            # contents[m] 和 styles[m] 分别是模态 m 对应的 Tensor 
            for m in range(4):
                # 均值池化降维，把整个 3D/2D Feature Map 压缩为一个一维全局表征向量
                # contents 的 shape 为 [B, 4, C, H, W, D]
                # styles 的 shape 为 [B, 4, C]
                c_feat = contents[:, m].view(contents.size(0), contents.size(2), -1).mean(-1).cpu().numpy()[0]
                s_feat = styles[:, m].view(styles.size(0), styles.size(2), -1).mean(-1).cpu().numpy()[0]
                
                all_contents[m].append(c_feat)
                all_styles[m].append(s_feat)

    # 3. 组织以供 TSNE 降维
    print("Running t-SNE reduction...")
    content_X, content_y, content_sample = [], [], []
    style_X, style_y, style_sample = [], [], []
    
    for m in range(4):
        content_X.extend(all_contents[m])
        content_y.extend([m]*len(all_contents[m]))
        content_sample.extend(list(range(len(all_contents[m]))))
        
        style_X.extend(all_styles[m])
        style_y.extend([m]*len(all_styles[m]))
        style_sample.extend(list(range(len(all_styles[m]))))

    content_X, content_y, content_sample = np.array(content_X), np.array(content_y), np.array(content_sample)
    style_X, style_y, style_sample = np.array(style_X), np.array(style_y), np.array(style_sample)

    tsne_model = TSNE(n_components=2, perplexity=min(10, num_samples-1), random_state=42)
    content_tsne = tsne_model.fit_transform(content_X)
    style_tsne = tsne_model.fit_transform(style_X)

    # 4. 绘图
    print("Plotting figures...")
    import matplotlib.cm as cm
    colors = cm.get_cmap('tab20', num_samples)
    markers = ['o', 's', '^', 'D', 'v', '<', '>', 'p', '*', 'h', 'H', 'd', 'P', 'X']
    
    # 建立图例用的 handles 列表
    from matplotlib.lines import Line2D
    # 序列用形状表示
    modality_handles = [Line2D([0], [0], marker=markers[m % len(markers)], color='w', label=modality_names[m], markerfacecolor='gray', markersize=8) for m in range(4)]
    # 病人用颜色表示
    patient_handles = [Line2D([0], [0], marker='o', color='w', label=f'Patient {s+1}', markerfacecolor=colors(s), markersize=8) for s in range(num_samples)]
    
    plt.figure(figsize=(16, 6))

    # 子图1 - 解剖表征 分布 (Anatomical Content)
    plt.subplot(1, 2, 1)
    for m in range(4):
        idx_m = (content_y == m)
        marker = markers[m % len(markers)]
        for s in range(num_samples):
            idx = idx_m & (content_sample == s)
            if np.any(idx):
                plt.scatter(content_tsne[idx, 0], content_tsne[idx, 1], color=colors(s), marker=marker, s=80, alpha=0.8, edgecolors='w')
    plt.title('t-SNE of Anatomical Representations (Contents)')
    plt.xlabel('t-SNE dim 1')
    plt.ylabel('t-SNE dim 2')
    plt.legend(handles=modality_handles + patient_handles, loc='best', fontsize='small', ncol=2)

    # 子图2 - 模态表征 分布 (Modality Styles)
    plt.subplot(1, 2, 2)
    for m in range(4):
        idx_m = (style_y == m)
        marker = markers[m % len(markers)]
        for s in range(num_samples):
            idx = idx_m & (style_sample == s)
            if np.any(idx):
                plt.scatter(style_tsne[idx, 0], style_tsne[idx, 1], color=colors(s), marker=marker, s=80, alpha=0.8, edgecolors='w')
    plt.title('t-SNE of Modality Representations (Styles)')
    plt.xlabel('t-SNE dim 1')
    plt.ylabel('t-SNE dim 2')
    plt.legend(handles=modality_handles + patient_handles, loc='best', fontsize='small', ncol=2)

    save_fig_path = '/root/dcreg/DC-Seg/distributions_tsne.png'
    plt.savefig(save_fig_path, dpi=300, bbox_inches='tight')
    plt.close()
    
    print(f"Visualization done! Successfully saved to {save_fig_path}")

if __name__ == '__main__':
    main()`

**Comment by CuCl-2** [comment]
那我晚上看看吧，现在不太方便，估计哪里搞错了。虽然我用的是brats2020，但我感觉2018也没啥差别

**Comment by Mike-ihr** [comment]
> 那我晚上看看吧，现在不太方便，估计哪里搞错了。虽然我用的是brats2020，但我感觉2018也没啥差别

好的非常感谢。因为我也在做序列解耦这方面的探索，我自己设计的对比实验，也是发现序列特征是比较好区分的，但是解剖特征想拉近实在是太难了。所以从您的实验过程来看，解剖结构对比loss降低幅度很低，但从可视化效果看起来是可以的嘛

**Comment by CuCl-2** [comment]
[plot_github_issue.ipynb](https://github.com/user-attachments/files/26651823/plot_github_issue.ipynb)

**Comment by CuCl-2** [comment]
你看看用这个代码试试呢，我找到了原始的画图代码，你看看。有后续也和我说一下

**Comment by Mike-ihr** [comment]
我使用github仓库提供的权重以及crop 112的设置并没有复现出聚类的效果。但是在BraTs2018数据集上采用crop 80我自己训练的权重复现出了聚类的效果。根本原因在于，在解剖特征做t-sne的时候，如果输入是[B,S,C,H,W,D]我的聚类方式是对所有空间求平均即结果是[B,S,C]，因此效果不好。你提供的脚本则是对通道进行求平均即[B,S,H*W*D]。但是这样操作是合理的吗？

**Comment by CuCl-2** [comment]
你这样岂不是丢失了空间信息吗，都被你平均掉了，那anatomical信息肯定丢失了呀

**Comment by CuCl-2** [comment]
你是要写paper吗，可以合作呀。vx: 15259198895
